File: Java/001_TwoSum.java
package leetCode;

import java.util.HashMap;
import java.util.Map;

public class twoNum {

	public int[] twoSum(int[] nums, int target) {
		int[] results = new int[2];
		Map<Integer, Integer> hash = new HashMap<Integer, Integer>();
		
		for(int i = 0; i < nums.length; i++) {
			
			int anOther = target - nums[i];
			
			if(hash.containsKey(anOther)) {
				results[0] = hash.get(anOther);
				results[1] = i;
			}			
			hash.put(nums[i], i);
		}		
		return results;
	}
	
	public static void main(String[] args) {
		
		twoNum t = new twoNum();
		int[] nums = new int[4];
		nums[0] = 2;
		nums[1] = 7;
		nums[2] = 11;
		nums[3] = 15;
		
		int target = 26;
		
		int[] res = new int[2];
		res = t.twoSum(nums, target);
		System.out.println(res[0]);
		System.out.println(res[1]);
	}
	
}

File: Cpp/001_TwoSum.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
using namespace std;

class Solution {
public:
    	vector<int> twoSum(vector<int> &numbers, int target) {
		
		vector<int> result;
		unordered_map<int, int> hash;

		for (int i = 0; i < numbers.size(); i++) {
			//find another nums
			int anOther = target - numbers[i];
			
			//Searches the container for an element 
			//with k as key and returns an iterator 
			//to it if found, otherwise it returns 
			//an iterator to unordered_map::end
			if (hash.find(anOther) != hash.end()) {
			    result.push_back(hash[anOther]);
				result.push_back(i);
			}

			//number was not found. Put it in the map.
			hash[numbers[i]] = i;
		}
		return result;
	}
};
